### PR TITLE
additional support for float8_e4m3fnuz and _e5m2fnuz

### PR DESCRIPTION
**Pull request by jeffdaily**
Follow up to #107586.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by jeffdaily** [comment]
CC @galexite for review.

File: aten/src/ATen/NumericUtils.h [hunk 1]
```diff
@@ -118,6 +136,14 @@ inline C10_HOST_DEVICE bool _isinf(at::Float8_e4m3fn val) {
   return false;
 }
 
+inline C10_HOST_DEVICE bool _isinf(at::Float8_e5m2fnuz val) {
+  return false;
```
Comment (umangyadav) [inline aten/src/ATen/NumericUtils.h:140]: I dont' have enough context but this type doesn't have infinity therefore it doesn't make sense to thave `__isinf` defined. 

https://github.com/pytorch/pytorch/blob/b8ce05456cd6e332746165e469ab1e9048c3daf5/c10/util/Float8_e4m3fnuz-inl.h#L41

https://en.cppreference.com/w/cpp/numeric/math/isinf 

std::isinf is only defined floating points which have infs and not other types.

File: aten/src/ATen/NumericUtils.h [hunk 2]
```diff
@@ -80,6 +82,22 @@ inline C10_HOST_DEVICE bool _isnan(T val) {
...
+        type = 0>
+inline C10_HOST_DEVICE bool _isnan(T val) {
+  return val.isnan();
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_same<T, at::Float8_e4m3fnuz>::value, int>::
+        type = 0>
+inline C10_HOST_DEVICE bool _isnan(T val) {
```
Comment (peterbell10) [inline aten/src/ATen/NumericUtils.h:97]:
```suggestion
inline C10_HOST_DEVICE bool _isnan(at::Float8_e4m3fnuz val) {
```

File: aten/src/ATen/NumericUtils.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/NumericUtils.h:140]: `_isinf` and `_isnan` are defined for all types including integers, regardless of whether they have special values. This makes generic code easier to write.

File: aten/src/ATen/NumericUtils.h [hunk 3]
```diff
@@ -80,6 +82,22 @@ inline C10_HOST_DEVICE bool _isnan(T val) {
   return val.isnan();
 }
 
+template <
+    typename T,
+    typename std::enable_if<std::is_same<T, at::Float8_e5m2fnuz>::value, int>::
+        type = 0>
+inline C10_HOST_DEVICE bool _isnan(T val) {
```
Comment (peterbell10) [inline aten/src/ATen/NumericUtils.h:89]:
```suggestion
inline C10_HOST_DEVICE bool _isnan(at::Float8_e5m2fnuz val) {
```

File: torch/storage.py [hunk 4]
```diff
@@ -1070,6 +1078,16 @@ def float8_e4m3fn(self):
         _warn_typed_storage_removal()
         return self._to(torch.float8_e4m3fn)
 
+    def float8_e5m2fnuz(self):
```
Comment (peterbell10) [inline torch/storage.py:1081]: Should we be adding API surface to deprecated `TypedStorage`?

File: aten/src/ATen/NumericUtils.h (hunk 2 above)
Comment (jeffdaily) [reply aten/src/ATen/NumericUtils.h:97]: Is there a reason why you suggest this change for the fnuz-suffixed type and no the other f8 types on lines 69-83?

Comment (peterbell10) [reply aten/src/ATen/NumericUtils.h:97]: I think every overload that does `is_same<T, concrete_type>` could be update, it's just this that these ones are new in this PR.

Comment (jeffdaily) [reply aten/src/ATen/NumericUtils.h:97]: Could that perhaps be a follow-up PR that addresses your concern for all types that do this?

**Comment by jeffdaily** [comment]
@peterbell10 you are the only reviewer thus far with comments (thank you).  Is it possible for your concerns to be addressed in a follow-up PR?  This PR aims to bring the f8 fnuz type support up to par with the other f8 type support.  Your concerns might be broader than just this current PR and would include the prior f8 type support PR.

**Review by peterbell10 — APPROVED** [review]

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 5]
```diff
@@ -838,7 +838,7 @@ _scaled_mm_out_cuda(const Tensor& mat1, const Tensor& mat2,
   at::native::resize_output(out, {mat1_sizes[0], mat2_sizes[1]});
   at::native::resize_output(amax, {});
 
-#if !defined(USE_ROCM) && !defined(_MSC_VER)
+#if !defined(USE_ROCM) && !defined(_MSC_VER) || (defined(USE_ROCM) && ROCM_VERSION >= 50700)
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/Blas.cpp:841]: scaled_gemm is not supported on ROCm so this was a mistake.  Will revert this change.  Thanks for the review and catching this.

[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/Blas.cpp:841]:
```suggestion
#if !defined(USE_ROCM) && !defined(_MSC_VER)
```

File: c10/util/Float8_e4m3fnuz-inl.h [hunk 6]
```diff
@@ -12,21 +13,208 @@ namespace c10 {
...
 }
 
 /// Special values helper
 
-C10_HOST_DEVICE inline bool Float8_e4m3fnuz::isnan() const {
+inline C10_HOST_DEVICE bool Float8_e4m3fnuz::isnan() const {
   return x == 0b10000000;
 }
 
+/// Arithmetic
```
Comment (vkuzo) [inline c10/util/Float8_e4m3fnuz-inl.h:31]: note: can we remove this? context: https://github.com/pytorch/pytorch/issues/113663, this should not have been added to the other float8 flavors.

File: c10/util/Float8_e4m3fnuz-inl.h [hunk 7]
```diff
@@ -12,21 +13,208 @@ namespace c10 {
...
+}
+
+inline C10_HOST_DEVICE Float8_e4m3fnuz& operator/=(
+    Float8_e4m3fnuz& a,
+    const Float8_e4m3fnuz& b) {
+  a = a / b;
+  return a;
+}
+
+/// Arithmetic with floats
```
Comment (vkuzo) [inline c10/util/Float8_e4m3fnuz-inl.h:86]: can we remove this? context: https://github.com/pytorch/pytorch/issues/113663, this should not have been added to the other float8 flavors.

File: c10/util/Float8_e4m3fnuz-inl.h (hunk 6 above)
Comment (vkuzo) [reply c10/util/Float8_e4m3fnuz-inl.h:31]: basically this entire section can be removed, unless this is needed for real use cases

File: torch/csrc/jit/tensorexpr/types.cpp [hunk 8]
```diff
@@ -74,8 +74,15 @@ int Dtype::byte_size() const {
     scalar_size = sizeof(Type); \
     break;
 
-    AT_FORALL_SCALAR_TYPES_AND5(
-        Bool, Half, BFloat16, Float8_e5m2, Float8_e4m3fn, TYPE_CASE);
+    AT_FORALL_SCALAR_TYPES_AND7(
```
Comment (malfet) [inline torch/csrc/jit/tensorexpr/types.cpp:77]: Why is this needed here? For ONNX tracing or something else? As TorchScript is in maintenance mode, I propose avoid adding new dtypes there

File: torch/_inductor/graph.py [hunk 9]
```diff
@@ -90,6 +90,8 @@ def supported_dtype_of_cpp_wrapper(dtype, cuda):
     if cuda:
         supported_dtype.add(torch.float8_e4m3fn)
         supported_dtype.add(torch.float8_e5m2)
+        supported_dtype.add(torch.float8_e4m3fnuz)
+        supported_dtype.add(torch.float8_e5m2fnuz)
```
Comment (malfet) [inline torch/_inductor/graph.py:94]: Should we add those only if `hip`? Or `cuda` is alias for `hip`?

File: c10/util/Float8_e5m2fnuz.h [hunk 10]
```diff
@@ -45,20 +44,274 @@ namespace detail {
...
+      -14336.0f,
+      -16384.0f,
+      -20480.0f,
+      -24576.0f,
+      -28672.0f,
+      -32768.0f,
+      -40960.0f,
+      -49152.0f,
+      -57344.0f,
+  };
```
Comment (malfet) [inline c10/util/Float8_e5m2fnuz.h:305]: This was part of the feedback for original PR: having a lookup table of the `constexpr inline` function will bloat binary size tremendously. If you need it it inlinable, let's figure out how it could be done differently

File: c10/util/Float8_e4m3fnuz.h [hunk 11]
```diff
@@ -70,7 +324,7 @@ C10_HOST_DEVICE inline uint8_t fp8e4m3fnuz_from_fp32_value(float f) {
 
   /*
    * A mask for converting fp32 numbers lower than fp8e4m3fnuz normal range
-   * into denormalized representation.
+   * into denorm representation
```
Comment (malfet) [inline c10/util/Float8_e4m3fnuz.h:327]: Why changing that?

File: c10/util/Float8_e5m2fnuz-inl.h [hunk 12]
```diff
@@ -12,21 +13,212 @@ namespace c10 {
...
+  a = a + b;
+  return a;
+}
+
+inline C10_HOST_DEVICE Float8_e5m2fnuz& operator-=(
+    Float8_e5m2fnuz& a,
+    const Float8_e5m2fnuz& b) {
+  a = a - b;
+  return a;
+}
```
Comment (malfet) [inline c10/util/Float8_e5m2fnuz-inl.h:74]: Hmm, let me think how one can refactor those repetitive `-inl` files, as they are identical to others barring type name. I would hate reintroducing TH pattern, but perhaps there are no way around it

**Review by malfet — CHANGES_REQUESTED** [review]
Mostly LGTM, but keeping 256-element lookup tables in the inline routine feels somewhat wrong. Let's discuss how it could be avoided

File: c10/util/Float8_e4m3fnuz.h (hunk 11 above)
Comment (jeffdaily) [reply c10/util/Float8_e4m3fnuz.h:327]: I was doing a side by side diff of the original 2 F8 headers versus the fnuz headers.  I found it useful to minimize the differences between the corresponding files (ex e4m3fn vs e4m3fnuz) including even the superficial differences, preferring the original 2 F8 type headers.

File: aten/src/ATen/AccumulateType.h [hunk 13]
```diff
@@ -87,6 +89,8 @@ MPS_ACC_TYPE(BFloat16, float);
 MPS_ACC_TYPE(Half, float);
 MPS_ACC_TYPE(Float8_e5m2, float);
 MPS_ACC_TYPE(Float8_e4m3fn, float);
+MPS_ACC_TYPE(Float8_e5m2fnuz, float);
+MPS_ACC_TYPE(Float8_e4m3fnuz, float);
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/AccumulateType.h:93]: This PR only brings the F8 fnuz types up to parity with the non-fnuz types.  If it was there for the original 2, I did it for the other 2.

File: torch/csrc/jit/tensorexpr/types.cpp (hunk 8 above)
Comment (jeffdaily) [reply torch/csrc/jit/tensorexpr/types.cpp:77]: Then why were the other 2 F8 types added?

File: torch/_inductor/graph.py (hunk 9 above)
Comment (jeffdaily) [reply torch/_inductor/graph.py:94]: What is the danger in adding them unconditionally?  fnuz is not just a hip thing.

File: c10/util/Float8_e5m2fnuz.h (hunk 10 above)
Comment (jeffdaily) [reply c10/util/Float8_e5m2fnuz.h:305]: I know it was part of the original feedback.  Perhaps if it was made static in addition to the constexpr inline?  Only constexpr static data members are implicitly inline.  W.r.t. bloat, the LUT is 1KB in size.  Would it only be inlined where used?  But the F8 types have very few places where they are exposed to kernel dispatch tables; based on other comments the arithmetic features of F8 types should be removed, reducing the potential bloat further.

File: c10/util/Float8_e5m2fnuz-inl.h (hunk 12 above)
Comment (jeffdaily) [reply c10/util/Float8_e5m2fnuz-inl.h:74]: I was struggling to avoid linking errors unless using the currently proposed implementation.

File: c10/util/Float8_e4m3fnuz-inl.h (hunk 6 above)
Comment (jeffdaily) [reply c10/util/Float8_e4m3fnuz-inl.h:31]: Should we remove the arithmetic from the other F8 flavors as part of this PR, or a follow-up PR?

File: c10/util/Float8_e5m2fnuz.h (hunk 10 above)
Comment (jeffdaily) [reply c10/util/Float8_e5m2fnuz.h]: No longer using a lookup table.  Switched to bit conversions.

**Comment by jeffdaily** [comment]
> Mostly LGTM, but keeping 256-element lookup tables in the inline routine feels somewhat wrong. Let's discuss how it could be avoided

Replaced lookup tables with bit shifts.

File: c10/util/Float8_e4m3fnuz-inl.h (hunk 6 above)
Comment (vkuzo) [reply c10/util/Float8_e4m3fnuz-inl.h:32]: I'd vote separate PR to keep things small

**Comment by PaliC** [comment]
@pytorchbot -m "breaking internal builds" -c "ghfirst"

Sorry got to send this back, we tend to care about strict aliasing more internally. We're getting the following errors on a bunch of builds as we treat the following warnings as errors

```
[pytorch]/c10/util/Float8_fnuz_cvt.h: In instantiation of ‘float c10::detail::fp8_fnuz_to_fp32_value(uint8_t) [with unsigned int we = 4; unsigned int wm = 3; uint8_t = unsigned char]’:
[pytorch]/c10/util/Float8_e4m3fnuz-inl.h:23:48:   required from here
[pytorch]/c10/util/Float8_fnuz_cvt.h:23:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   23 |     return reinterpret_cast<const float&>(ifNaN);
      |                                         ^~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h:53:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   53 |   return reinterpret_cast<const float&>(retval);
      |                                         ^~~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h: In instantiation of ‘float c10::detail::fp8_fnuz_to_fp32_value(uint8_t) [with unsigned int we = 5; unsigned int wm = 2; uint8_t = unsigned char]’:
[pytorch]/c10/util/Float8_e5m2fnuz-inl.h:23:48:   required from here
[pytorch]/c10/util/Float8_fnuz_cvt.h:23:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   23 |     return reinterpret_cast<const float&>(ifNaN);
      |                                         ^~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h:53:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   53 |   return reinterpret_cast<const float&>(retval);
      |                                         ^~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by PaliC** [comment]
oops 
@pytorchbot revert -m "breaking internal builds" -c "ghfirst"

**Comment by PaliC** [comment]
@pytorchbot revert -m "breaking internal builds" -c "ghfirst"

Sorry got to send this back, we tend to care about strict aliasing more internally. We're getting the following errors on a bunch of builds as we treat the following warnings as errors

```
[pytorch]/c10/util/Float8_fnuz_cvt.h: In instantiation of ‘float c10::detail::fp8_fnuz_to_fp32_value(uint8_t) [with unsigned int we = 4; unsigned int wm = 3; uint8_t = unsigned char]’:
[pytorch]/c10/util/Float8_e4m3fnuz-inl.h:23:48:   required from here
[pytorch]/c10/util/Float8_fnuz_cvt.h:23:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   23 |     return reinterpret_cast<const float&>(ifNaN);
      |                                         ^~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h:53:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   53 |   return reinterpret_cast<const float&>(retval);
      |                                         ^~~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h: In instantiation of ‘float c10::detail::fp8_fnuz_to_fp32_value(uint8_t) [with unsigned int we = 5; unsigned int wm = 2; uint8_t = unsigned char]’:
[pytorch]/c10/util/Float8_e5m2fnuz-inl.h:23:48:   required from here
[pytorch]/c10/util/Float8_fnuz_cvt.h:23:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   23 |     return reinterpret_cast<const float&>(ifNaN);
      |                                         ^~~~~
[pytorch]/c10/util/Float8_fnuz_cvt.h:53:41: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   53 |   return reinterpret_cast<const float&>(retval);
      |                                         ^~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by jeffdaily** [comment]
@PaliC I think I have a solution pushed now.  Do you need to try and do an internal import before landing?

**Review by malfet — APPROVED** [review]

**Comment by jeffdaily** [comment]
@pytorch merge -f "relanding after fixing internal build; unrelated test failures"

**Comment by anupambhatnagar** [comment]
I'm using omnitrace for system level profiling on MI300 and need to build pytorch from source. This commit is breaking build from source. I had the build working before the weekend. Here's the error I get:

```
[1/581] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/torch_hip_generated_SparseMatMul.hip.o
FAILED: caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/torch_hip_generated_SparseMatMul.hip.o /home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/torch_hip_generated_SparseMatMul.hip.o
cd /home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip && /home/anupamb/miniconda3/envs/pt-install-test2/lib/python3.12/site-packages/cmake/data/bin/cmake -E make_directory /home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/. && /home/anupamb/miniconda3/envs/pt-install-test2/lib/python3.12/site-packages/cmake/data/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/./torch_hip_generated_SparseMatMul.hip.o -P /home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/torch_hip_generated_SparseMatMul.hip.o.cmake
In file included from /home/anupamb/pytorch/aten/src/ATen/native/sparse/hip/SparseMatMul.hip:28:
/home/anupamb/pytorch/aten/src/ATen/hip/HIPDataType.h:99:14: error: use of undeclared identifier 'HIP_R_8F_E4M3_FNUZ'
      return HIP_R_8F_E4M3_FNUZ;
             ^
/home/anupamb/pytorch/aten/src/ATen/hip/HIPDataType.h:101:14: error: use of undeclared identifier 'HIP_R_8F_E5M2_FNUZ'
      return HIP_R_8F_E5M2_FNUZ;
             ^
2 errors generated when compiling for gfx942.
CMake Error at torch_hip_generated_SparseMatMul.hip.o.cmake:200 (message):
  Error generating file
  /home/anupamb/pytorch/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/sparse/hip/./torch_hip_generated_SparseMatMul.hip.o
```

The offending code block seems to be https://github.com/pytorch/pytorch/blob/b90199935078b509c4a721893ea7d478878e7734/aten/src/ATen/cuda/CUDADataType.h#L96-L100. This landed earlier today. 

@jeffdaily @malfet @xw285cornell

**Comment by anupambhatnagar** [comment]
I've tried this with rocm 6.0 and 6.1

**Comment by jeffdaily** [comment]
@anupambhatnagar why did you reopen this PR?  It was already merged.  You can comment on this PR or file an issue but opening the PR doesn't revert it, and this PR was building correctly with the official ROCm 6.0 release.  Note that ROCm 6.1 is not released yet so you must be using some internal preview build we provide to customers.  Let me find a suitable work-around for these unreleased builds of ROCm.

**Comment by jeffdaily** [comment]
@anupambhatnagar what about this proposed fix?

https://github.com/pytorch/pytorch/pull/118137

**Comment by anupambhatnagar** [comment]
@jeffdaily sorry about re-opening this. I wasn't completely aware of the implication of re-opening the PR. This commit broke the ROCM pytorch build for us. I have tested your patch in #118137 with rocm-6.0 on MI300 and it works if the the correct LD_LIBRARY_PATH is set in the environment. Is setting the LD_LIBRARY_PATH a requirement with rocm builds? I don't recall seeing that in the official pytorch install documentation.
